### PR TITLE
0.2/681 leaving function

### PR DESCRIPTION
**Pull request by charlesstrange2**
Prospective PR -  I will add tests shortly
@alexalala Have a look at the colour choices for disabled families on search.

@katjam The leave form turned out not to need a checkbox. This simplified the controller; I also pulled the validation rules into a formRequest and the Gate just wasnt working - the route is, however, in the web:auth route group which will protect it against unauthenticated hits, and all authenticated users are authorised to update families.

### CODE REVIEW

File: resources/views/service/index_registration.blade.php [hunk 1]
```diff
@@ -37,7 +37,9 @@
                             <td class="center">{{ $registration->family->entitlement }}</td>
                             <td class="center">{{ $registration->family->rvid }}</td>
                             <td>
-                            <button onclick="window.location.href='{{ URL::route('service.registration.edit', ['id' => $registration->id ]) }}'">Select</button>
+                            <button onclick="window.location.href='{{ URL::route('service.registration.edit', ['id' => $registration->id ]) }}'"
+                                @if( isset($registration->family->leaving_on) ) disabled @endif
+                            > Select</button>
```
Comment (katjam) [inline resources/views/service/index_registration.blade.php:42]: How would you turn them back on? Or is that not something we'd ever want to need?

File: app/Http/Requests/StoreUpdateRegistrationRequest.php [hunk 2]
```diff
@@ -15,7 +16,9 @@ class StoreUpdateRegistrationRequest extends FormRequest
     public function authorize()
     {
         // Todo: replace with check that user can make this request rather than wave them through.
-        return true;
+        // Refuse updates to off-scheme users.
+        $registration = Registration::find($this->route('registration'))->first();
+        return (!isset($registration->family->leaving_on));
```
Comment (katjam) [inline app/Http/Requests/StoreUpdateRegistrationRequest.php:21]: Good idea.

File: routes/web.php [hunk 3]
```diff
@@ -65,8 +65,16 @@
...
+
+    // Update (deactivate) a Registration's Family
+    Route::put('/registrations/{registration}/family', [
+        'as' => 'service.registration.family',
+        'uses' => 'Service\FamilyController@update',
+    ]);
+
     // Printables
 
+
```
Comment (katjam) [inline routes/web.php]: intentional extra line?

File: tests/Feature/EditPageTest.php [hunk 4]
```diff
@@ -380,4 +380,28 @@ public function itShowsTheLoggedInUserDetails()
         ;
     }
 
+    /** @test */
+    public function itShowsTheLeavingFormIfFamilyIsOnScheme()
+    {
+    }
+
+    /** @test */
+    public function itDoeNotShowTheLeavingFormIfFamilyIsOnScheme()
```
Comment (katjam) [inline tests/Feature/EditPageTest.php]: ifFamilyHasLeftScheme ?

File: tests/Feature/EditPageTest.php [hunk 5]
```diff
@@ -380,4 +380,28 @@ public function itShowsTheLoggedInUserDetails()
...
+
+    /** @test */
+    public function itWillRejectLeavingWithoutAReason()
+    {
+    }
+
+    /** @test */
+    public function itWillAcceptLeavingWithAReason()
+    {
+    }
```
Comment (katjam) [inline tests/Feature/EditPageTest.php:436]: +1 more. It will not accept an invalid leaving reason

File: tests/Unit/Routes/WebRoutesTest.php [hunk 6]
```diff
@@ -210,4 +210,9 @@ public function testCentreRegistrationsSummaryGate()
             ->assertResponseOK()
         ;
     }
+
+    /** @test */
+    public function testRegistrationFamilyUpdateGate()
```
Comment (katjam) [inline tests/Unit/Routes/WebRoutesTest.php:215]: This just returns true right now, correct?

**Review by katjam — CHANGES_REQUESTED** [review]
Only small ones. Looks pretty good.

File: resources/views/service/index_registration.blade.php (hunk 1 above)
Comment (charlesstrange2) [reply resources/views/service/index_registration.blade.php:42]: When re-enable people becomes a thing, then we'll do that. At the moment they'll probably just reregister them as a second family record.

File: tests/Unit/Routes/WebRoutesTest.php (hunk 6 above)
Comment (charlesstrange2) [reply tests/Unit/Routes/WebRoutesTest.php:215]: An unauthed user will be blocked blocked. An auth'd user will be waved through.

File: public/css/main.css [hunk 7]
```diff
@@ -130,6 +130,12 @@ button {
 	box-shadow:0 3px 12px -1px rgba(0,0,0,.2);
 }
 
+button:disabled,
+button[disabled] {
```
Comment (alexalala) [inline public/css/main.css]: you shouldn't need the button[disabled] line as well as the :disabled selector! Unless my CSS is bad and you needed to override something haha

**Review by alexalala — APPROVED** [review]
tiny CSS change, 
also you could get the family's name on the info message 'Family 2 de-activated.' rather than Family 2.

Other than that it looks good 👍

File: resources/views/service/index_registration.blade.php (hunk 1 above)
Comment (katjam) [reply resources/views/service/index_registration.blade.php:42]: Fair.

File: tests/Unit/Routes/WebRoutesTest.php (hunk 6 above)
Comment (katjam) [reply tests/Unit/Routes/WebRoutesTest.php:215]: Of course.

**Review by katjam — APPROVED** [review]
Let's get it in. I will carry on with the Search and Routes tests in a new branch.

:paintbrush:

**Pull request merged**